File: web/src/features/terminal/lib/host-theme.ts
import { apiFetch } from '@/lib/api'
import { themeRegistry } from '@/extensions/themes/theme-registry'
import { readTerminalThemePayload } from '../hooks/use-terminal-theme'

/**
 * Tell the daemon what the host terminal's default colours are, so every PTY it spawns
 * AFTERWARDS is born answering an OSC 10/11 query with them.
 *
 * This is the half of theme propagation `terminalSetTheme` structurally cannot do. That one
 * travels down a session's own socket, so it can only ever reach a session that already
 * exists — and a vendor CLI is exec'd by the daemon at session-creation time, before any
 * client can attach. A CLI that reads the background once at startup has therefore already
 * asked, and been told x/vt's hardcoded black, by the time the first `terminalSetTheme`
 * lands. Claude Code hid that for a long time because it also subscribes to DEC 2031 and
 * re-queries when the late push notifies it; codex 0.146.0 has no 2031 support at all, so
 * the answer it got at birth was the only one it would ever get, and a light-mode Crowbar
 * came up with a dark Codex.
 *
 * Best-effort by design — a failed push leaves the daemon on its previous value, which is
 * exactly today's behaviour, never a wrong one.
 */
export async function pushHostTerminalTheme(): Promise<boolean> {
  const { background, foreground } = readTerminalThemePayload()
  try {
    await apiFetch('/v0/settings/terminal/theme', {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ bg: background, fg: foreground }),
    })
    return true
  } catch {
    return false
  }
}

/**
 * Boot retry schedule, in ms. The push is the one daemon call that must land BEFORE the
 * user does anything, and it is a PUT — so `apiFetch`'s built-in retry (idempotent reads
 * only) does not cover it, and a daemon still coming up at boot would otherwise leave the
 * host theme unknown until the user happened to switch themes or open a terminal.
 *
 * Bounded and abandoned on the first success: this is a startup handshake, not a poll.
 */
const BOOT_RETRY_DELAYS_MS = [250, 500, 1000, 2000, 4000]

/**
 * Push the host theme now, and again on every theme change, for the lifetime of the app.
 * Returns a teardown for HMR disposal.
 */
export function startHostThemeSync(): () => void {
  let disposed = false
  let retryTimer: ReturnType<typeof setTimeout> | null = null

  const clearRetry = () => {
    if (retryTimer !== null) {
      clearTimeout(retryTimer)
      retryTimer = null
    }
  }

  // A theme change supersedes any in-flight boot retry: it pushes the newer value itself,
  // so letting the old schedule run on would only re-send a stale colour pair.
  const pushNow = () => {
    clearRetry()
    void pushHostTerminalTheme()
  }

  const pushWithBootRetry = (attempt: number) => {
    void pushHostTerminalTheme().then((ok) => {
      if (ok || disposed || attempt >= BOOT_RETRY_DELAYS_MS.length) return
      retryTimer = setTimeout(() => {
        retryTimer = null
        pushWithBootRetry(attempt + 1)
      }, BOOT_RETRY_DELAYS_MS[attempt])
    })
  }

  pushWithBootRetry(0)
  const unlisten = themeRegistry.onThemeChange(pushNow)

  return () => {
    disposed = true
    clearRetry()
    unlisten()
  }
}

File: api/internal/api/v0/endpoints/terminal/handlers/hosttheme.go
package handlers

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"github.com/char2cs/crowbar/api/internal/api/libs"
	engineterminal "github.com/char2cs/crowbar/api/internal/engine/terminal"
)

// hostThemeReq is the host terminal's resolved default colours, in the same "#rrggbb" wire
// form the per-session theme frame already carries (resolve-css-color.ts output).
//
// There is no `dark` field, and that omission is deliberate rather than an oversight: the
// polarity flag exists only to choose which DEC 2031 CSI ?997;n report to inject into an
// ALREADY-RUNNING subscribed app, which is the per-session channel's job. Nothing born from
// this endpoint has a foreground app yet, so a polarity here would be a field with no reader.
type hostThemeReq struct {
	Bg string `json:"bg"`
	Fg string `json:"fg"`
}

// SetHostTheme handles PUT /v0/settings/terminal/theme: it records the host terminal's
// light/dark colours as what every SUBSEQUENTLY SPAWNED PTY answers an OSC 10/11 query with.
//
// This is the half of theme propagation the per-session channel structurally cannot cover.
// A vendor CLI is exec'd by the daemon at session-creation time, so it is already running —
// and, if it detects light/dark by querying the background, has usually already asked —
// before any client can attach and push a theme down the session's socket. Claude Code
// survived that ordering only because it also subscribes to DEC 2031 and re-queries when the
// late push notifies it; codex 0.146.0 has no 2031 support, so the answer it got at birth was
// final. This endpoint is what makes that first answer correct.
//
// It is idempotent and unconditional — the frontend pushes at boot and on every theme change,
// and the newest push simply wins.
func (h *Handlers) SetHostTheme(ctx *gin.Context) {
	var req hostThemeReq
	if err := ctx.ShouldBindJSON(&req); err != nil {
		libs.WriteErr(ctx, http.StatusBadRequest, err.Error())
		return
	}

	bg, fg := engineterminal.ParseHexColor(req.Bg), engineterminal.ParseHexColor(req.Fg)
	if bg == nil && fg == nil {
		// Neither channel parsed. Refusing is the honest answer: accepting would report
		// success for a call that changed nothing, and the caller would have no way to tell
		// a typo'd colour from a stored one.
		libs.WriteErr(ctx, http.StatusBadRequest,
			`theme requires at least one parseable colour ("#rgb", "#rrggbb" or "#rrggbbaa")`)
		return
	}
	h.termEng.SetHostTheme(bg, fg)

	ctx.Status(http.StatusNoContent)
}
